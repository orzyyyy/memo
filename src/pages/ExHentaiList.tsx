import React, { useState, useEffect } from 'react';
import { ExHentaiInfoItem } from '../../server/controller/ExhentaiController';
import './css/exhentai-list.css';

export interface DownloadProps {
  url: string;
  name?: string;
}
export interface ExHentaiListProps {
  dataSource?: ExHentaiInfoItem[];
  onDownload: ({ url, name }: DownloadProps) => void;
  wrapperHeight: number;
  onDetail: (url: string) => void;
}

const patch = 15;

const ExHentaiList = ({ dataSource = [], onDownload, onDetail }: ExHentaiListProps) => {
  const [realImgInfo, setRealImgInfo] = useState([] as (ExHentaiInfoItem & { height: number; width: number })[]);

  useEffect(() => {
    const getImgHeight = async (data: ExHentaiInfoItem[]) => {
      const target = [];
      for (const item of data) {
        const img = new Image();
        img.src = item.thumbnailUrl;
        const params: { width: number; height: number } = await new Promise(resolve => {
          img.onload = () => {
            resolve({ width: img.width, height: img.height });
          };
        });
        target.push({ ...params, ...item });
      }
      return target;
    };

    const updatePatchImg = async (data: ExHentaiInfoItem[], result: any[], patch: number) => {
      const target = data.splice(0, patch);
      result = [...result, ...(await getImgHeight(target))];
      setRealImgInfo(result);
      if (data.length) {
        updatePatchImg(data, result, patch);
      }
    };

    updatePatchImg(dataSource, [], patch);
  }, [dataSource]);

  return (
    <ul className="exhentai-list">
      {realImgInfo.map(item => {
        return (
          <li key={item.detailUrl}>
            <img
              alt={item.name}
              src={item.thumbnailUrl}
              style={{ height: item.height, width: item.width }}
              onClick={() => onDetail(item.detailUrl)}
              onContextMenu={() => onDownload({ url: item.detailUrl })}
            />
          </li>
        );
      })}
    </ul>
  );
};

export default ExHentaiList;
