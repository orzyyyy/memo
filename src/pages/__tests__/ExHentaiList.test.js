import React from 'react';
import { shallow } from 'enzyme';
import ExHentaiList from '../ExHentaiList';

describe('ExHentaiList', () => {
  const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  afterAll(() => {
    errorSpy.mockRestore();
  });

  it('render correctly', () => {
    const wrapper = shallow(
      <ExHentaiList
        dataSource={[
          {
            name:
              '[Zakotsu] Amanita Pantherina (COMIC Koh 2018-06) [Chinese] [嗷呜个人汉化] [Digital]',
            detailUrl: 'https://exhentai.org/g/1430261/17651c12eb/',
            postTime: 1559412420000,
            thumbnailUrl:
              'https://exhentai.org/t/5b/47/5b47c48abf0fb6c685869520cdb8ab0d97990e4d-1074410-1280-1807-jpg_250.jpg',
          },
        ]}
        wrapperHeight={1000}
        onDownload={() => {}}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
