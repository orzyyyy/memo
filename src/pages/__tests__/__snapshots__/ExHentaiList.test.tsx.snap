// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExHentaiList render correctly 1`] = `
<ul
  className="exhentai-list"
>
  <li
    key="https://exhentai.org/g/1430261/17651c12eb/-1559412420000"
  >
    <LazyLoad
      height={1000}
      offset={0}
      once={true}
      overflow={false}
      resize={false}
      scroll={true}
      scrollContainer=".main-page-content-wrapper"
      unmountIfInvisible={false}
    >
      <div
        onClick={[Function]}
        onContextMenu={[Function]}
        style={
          Object {
            "height": 500,
          }
        }
      >
        <img
          alt="[Zakotsu] Amanita Pantherina (COMIC Koh 2018-06) [Chinese] [嗷呜个人汉化] [Digital]"
          src="https://exhentai.org/t/5b/47/5b47c48abf0fb6c685869520cdb8ab0d97990e4d-1074410-1280-1807-jpg_250.jpg"
          style={
            Object {
              "height": "100%",
            }
          }
        />
      </div>
    </LazyLoad>
  </li>
</ul>
`;
